File: src/validation/validate.ts
import { isGitAvailable } from './isGitAvailable';
import { getUntrackedChanges } from 'workspace-tools';
import { isValidAuthType } from './isValidAuthType';
import { isValidChangeType } from './isValidChangeType';
import { isChangeFileNeeded } from './isChangeFileNeeded';
import { isValidGroupedPackageOptions, isValidGroupOptions } from './isValidGroupOptions';
import type { BeachballOptions } from '../types/BeachballOptions';
import { isValidChangelogOptions } from './isValidChangelogOptions';
import { readChangeFiles } from '../changefile/readChangeFiles';
import { getPackageInfos } from '../monorepo/getPackageInfos';
import { getPackageGroups } from '../monorepo/getPackageGroups';
import { getDisallowedChangeTypes } from '../changefile/getDisallowedChangeTypes';
import { areChangeFilesDeleted } from './areChangeFilesDeleted';
import { validatePackageDependencies } from '../publish/validatePackageDependencies';
import { gatherBumpInfo } from '../bump/gatherBumpInfo';
import { isValidDependentChangeType } from './isValidDependentChangeType';
import { getPackagesToPublish } from '../publish/getPackagesToPublish';
import { env } from '../env';
import type { PackageInfos } from '../types/PackageInfo';
import { bulletedList } from '../logging/bulletedList';

type ValidationOptions = {
  /**
   * If true, check whether change files are needed (and whether change files are deleted).
   */
  checkChangeNeeded?: boolean;
  /**
   * If true, don't error if change files are needed (just return isChangeNeeded true).
   */
  allowMissingChangeFiles?: boolean;
  /**
   * If true, validate that the dependencies of any packages with change files are valid
   * (not private).
   */
  checkDependencies?: boolean;
};

type ValidationResult = {
  /** True if change files are needed. Always false if `validateOptions.checkChangeNeeded` wasn't true. */
  isChangeNeeded: boolean;
};

/**
 * Validate configuration and exit 1 if it's invalid.
 * If `validateOptions.checkChangeNeeded` is true, also check whether change files are needed.
 */
export function validate(
  options: BeachballOptions,
  validateOptions: ValidationOptions,
  packageInfos: PackageInfos
): ValidationResult;
/** @deprecated Must provide the package infos */
export function validate(options: BeachballOptions, validateOptions?: ValidationOptions): ValidationResult;
export function validate(
  options: BeachballOptions,
  validateOptions?: ValidationOptions,
  packageInfos?: PackageInfos
): ValidationResult {
  const { allowMissingChangeFiles, checkChangeNeeded, checkDependencies } = validateOptions || {};

  console.log('\nValidating options and change files...');

  // Run the validation checks in stages and wait to exit until the end of the stage.
  // This provides more potentially useful info the user rather than hiding errors.
  let hasError = false;

  const logValidationError = (message: string) => {
    console.error(`ERROR: ${message}`);
    hasError = true;
  };

  if (!isGitAvailable(options.path)) {
    logValidationError('Please make sure git is installed and initialize the repository with "git init"');
    process.exit(1);
  }

  const untracked = getUntrackedChanges({ cwd: options.path });

  if (untracked.length) {
    console.warn('WARN: There are untracked changes in your repository:\n' + bulletedList(untracked));
    !env.isCI && console.warn('Changes in these files will not trigger a prompt for change descriptions');
  }

  packageInfos ||= getPackageInfos(options.path);

  if (options.all && options.package) {
    logValidationError('Cannot specify both "all" and "package" options');
  } else if (typeof options.package === 'string' && !packageInfos[options.package]) {
    logValidationError(`package "${options.package}" was not found`);
  } else {
    const invalidPackages = Array.isArray(options.package)
      ? options.package.filter(pkg => !packageInfos![pkg])
      : undefined;
    if (invalidPackages?.length) {
      logValidationError(`package(s) ${invalidPackages.map(pkg => `"${pkg}"`).join(', ')} were not found`);
    }
  }

  if (options.authType && !isValidAuthType(options.authType)) {
    logValidationError(`authType "${options.authType}" is not valid`);
  }

  if (options.command === 'publish' && options.token !== undefined) {
    if (options.token === '') {
      logValidationError(
        'token should not be an empty string. This usually indicates an incorrect variable name ' +
          'or forgetting to pass a secret into a workflow step.'
      );
    } else if (options.token.startsWith('$') && options.authType !== 'password') {
      logValidationError(
        `token appears to be a variable reference: "${options.token}" -- please check your workflow configuration.`
      );
    }
  }

  if (options.dependentChangeType && !isValidChangeType(options.dependentChangeType)) {
    logValidationError(`dependentChangeType "${options.dependentChangeType}" is not valid`);
  }

  if (options.type && !isValidChangeType(options.type)) {
    logValidationError(`Change type "${options.type}" is not valid`);
  }

  if (options.changelog && !isValidChangelogOptions(options.changelog)) {
    hasError = true; // the helper logs this
  }

  if (options.groups && !isValidGroupOptions(options.groups)) {
    hasError = true; // the helper logs this
  }

  // this exits the process if any package belongs to multiple groups
  const packageGroups = getPackageGroups(packageInfos, options.path, options.groups);

  if (options.groups && !isValidGroupedPackageOptions(packageInfos, packageGroups)) {
    hasError = true; // the helper logs this
  }

  const changeSet = readChangeFiles(options, packageInfos);

  for (const { changeFile, change } of changeSet) {
    const disallowedChangeTypes = getDisallowedChangeTypes(change.packageName, packageInfos, packageGroups);

    if (!change.type) {
      logValidationError(`Change type is missing in ${changeFile}`);
      hasError = true;
    } else if (!isValidChangeType(change.type)) {
      logValidationError(`Invalid change type detected in ${changeFile}: "${change.type}"`);
      hasError = true;
    } else if (disallowedChangeTypes?.includes(change.type)) {
      logValidationError(`Disallowed change type detected in ${changeFile}: "${change.type}"`);
      hasError = true;
    }

    if (!change.dependentChangeType) {
      logValidationError(`dependentChangeType is missing in ${changeFile}`);
      hasError = true;
    } else if (!isValidDependentChangeType(change.dependentChangeType, disallowedChangeTypes)) {
      logValidationError(`Invalid dependentChangeType detected in ${changeFile}: "${change.dependentChangeType}"`);
      hasError = true;
    }
  }

  if (hasError) {
    // If any of the above basic checks failed, it doesn't make sense to check if change files are needed
    process.exit(1);
  }

  let isChangeNeeded = false;

  if (checkChangeNeeded) {
    isChangeNeeded = isChangeFileNeeded(options, packageInfos);

    if (isChangeNeeded && !allowMissingChangeFiles) {
      logValidationError('Change files are needed!');
      console.log(options.changehint);
      process.exit(1); // exit here (this is the main poin)
    }

    if (options.disallowDeletedChangeFiles && areChangeFilesDeleted(options)) {
      logValidationError('Change files must not be deleted!');
      process.exit(1);
    }
  }

  if (!isChangeNeeded && checkDependencies && changeSet.length) {
    console.log('\nValidating package dependencies...');
    // Clone to avoid mutating shared package info during validation.
    const packageInfosForValidation = JSON.parse(JSON.stringify(packageInfos));
    // TODO: It would be preferable if this could be done without getting the full bump info,
    // or at least if the bump info could be passed back out to other methods which currently
    // duplicate the calculation (it can be expensive, especially in large repos).
    const bumpInfo = gatherBumpInfo(options, packageInfosForValidation);
    const packagesToPublish = getPackagesToPublish(bumpInfo, true /*validationMode*/);
    if (!validatePackageDependencies(packagesToPublish, bumpInfo.packageInfos)) {
      logValidationError(`One or more published packages depend on an unpublished package!

Consider one of the following solutions:
- If the unpublished package should be published, remove \`"private": true\` from its package.json.
- If it should NOT be published, verify that it is only listed under devDependencies of published packages.
`);
      process.exit(1);
    }
  }

  console.log();

  return {
    isChangeNeeded,
  };
}

File: rust/src/validation/validate.rs
use anyhow::Result;

use crate::changefile::change_types::get_disallowed_change_types;
use crate::changefile::changed_packages::get_changed_packages;
use crate::changefile::read_change_files::read_change_files;
use crate::git::commands::get_untracked_changes;
use crate::logging::bulleted_list;
use crate::monorepo::package_groups::get_package_groups;
use crate::monorepo::package_infos::get_package_infos;
use crate::monorepo::scoped_packages::get_scoped_packages;
use crate::types::change_info::ChangeSet;
use crate::types::options::ParsedOptions;
use crate::types::package_info::{PackageGroups, PackageInfos, ScopedPackages};
use crate::validation::are_change_files_deleted::are_change_files_deleted;
use crate::validation::validators::*;

pub struct ValidateOptions {
    pub check_change_needed: bool,
    pub allow_missing_change_files: bool,
    pub check_dependencies: bool,
}

impl Default for ValidateOptions {
    fn default() -> Self {
        Self {
            check_change_needed: false,
            allow_missing_change_files: false,
            check_dependencies: false,
        }
    }
}

#[derive(Debug)]
pub struct ValidationResult {
    pub is_change_needed: bool,
    pub package_infos: PackageInfos,
    pub package_groups: PackageGroups,
    pub scoped_packages: ScopedPackages,
    pub change_set: ChangeSet,
    pub changed_packages: Option<Vec<String>>,
}

/// Validation error that indicates the process should exit with code 1.
#[derive(Debug)]
pub struct ValidationError {
    pub message: String,
}

impl std::fmt::Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for ValidationError {}

/// Log a validation error and set the flag.
fn log_validation_error(msg: &str, has_error: &mut bool) {
    eprintln!("ERROR: {msg}");
    *has_error = true;
}

/// Run validation of options, change files, and packages.
pub fn validate(
    parsed: &ParsedOptions,
    validate_options: &ValidateOptions,
) -> Result<ValidationResult> {
    let options = &parsed.options;

    println!("\nValidating options and change files...");

    let mut has_error = false;

    // Check for untracked changes
    let untracked = get_untracked_changes(&options.path).unwrap_or_default();
    if !untracked.is_empty() {
        eprintln!(
            "WARN: There are untracked changes in your repository:\n{}",
            bulleted_list(&untracked.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        );
    }

    // Get package infos
    let package_infos = get_package_infos(options)?;

    // Validate --all and --package conflict
    if options.all && options.package.is_some() {
        log_validation_error(
            "Cannot specify both \"all\" and \"package\" options",
            &mut has_error,
        );
    } else if let Some(ref packages) = options.package {
        let mut invalid_reasons: Vec<String> = Vec::new();
        for pkg in packages {
            if !package_infos.contains_key(pkg) {
                invalid_reasons.push(format!("\"{pkg}\" was not found"));
            } else if package_infos[pkg].private {
                invalid_reasons.push(format!("\"{pkg}\" is marked as private"));
            }
        }
        if !invalid_reasons.is_empty() {
            log_validation_error(
                &format!(
                    "Invalid package(s) specified:\n{}",
                    bulleted_list(
                        &invalid_reasons
                            .iter()
                            .map(|s| s.as_str())
                            .collect::<Vec<_>>()
                    )
                ),
                &mut has_error,
            );
        }
    }

    // Validate auth type
    if let Some(ref auth_type) = options.auth_type {
        if !is_valid_auth_type(auth_type) {
            log_validation_error(
                &format!("authType \"{auth_type}\" is not valid"),
                &mut has_error,
            );
        }
    }

    // Validate dependent change type
    if let Some(dct) = options.dependent_change_type {
        if !is_valid_change_type_value(dct) {
            log_validation_error(
                &format!("dependentChangeType \"{dct}\" is not valid"),
                &mut has_error,
            );
        }
    }

    // Validate change type
    if let Some(ct) = options.change_type {
        if !is_valid_change_type_value(ct) {
            log_validation_error(
                &format!("Change type \"{ct}\" is not valid"),
                &mut has_error,
            );
        }
    }

    // Validate group options
    if let Some(ref groups) = options.groups {
        if !is_valid_group_options(groups) {
            has_error = true;
        }
    }

    // Get package groups
    let package_groups = get_package_groups(&package_infos, &options.path, &options.groups)?;

    // Validate grouped package options
    if options.groups.is_some()
        && !is_valid_grouped_package_options(&package_infos, &package_groups)
    {
        has_error = true;
    }

    let scoped_packages = get_scoped_packages(options, &package_infos);
    let change_set = read_change_files(options, &package_infos, &scoped_packages);

    // Validate each change file
    for entry in &change_set {
        let disallowed = get_disallowed_change_types(
            &entry.change.package_name,
            &package_infos,
            &package_groups,
            &options.disallowed_change_types,
        );

        let ct_str = entry.change.change_type.to_string();
        if !is_valid_change_type(&ct_str) {
            log_validation_error(
                &format!(
                    "Invalid change type detected in {}: \"{}\"",
                    entry.change_file, ct_str
                ),
                &mut has_error,
            );
        } else if let Some(ref disallowed) = disallowed {
            if disallowed.contains(&entry.change.change_type) {
                log_validation_error(
                    &format!(
                        "Disallowed change type detected in {}: \"{}\"",
                        entry.change_file, ct_str
                    ),
                    &mut has_error,
                );
            }
        }

        let dct_str = entry.change.dependent_change_type.to_string();
        if !is_valid_dependent_change_type(entry.change.dependent_change_type, &disallowed) {
            log_validation_error(
                &format!(
                    "Invalid dependentChangeType detected in {}: \"{}\"",
                    entry.change_file, dct_str
                ),
                &mut has_error,
            );
        }
    }

    if has_error {
        return Err(ValidationError {
            message: "Validation failed".to_string(),
        }
        .into());
    }

    // Check if change files are needed
    let mut is_change_needed = false;
    let mut changed_packages: Option<Vec<String>> = None;

    if validate_options.check_change_needed {
        let pkgs = get_changed_packages(options, &package_infos, &scoped_packages)?;
        is_change_needed = !pkgs.is_empty();

        if is_change_needed {
            let message = if options.all {
                "Considering the following packages due to --all"
            } else if options.package.is_some() {
                "Considering the specific --package"
            } else {
                "Found changes in the following packages"
            };
            let mut sorted = pkgs.clone();
            sorted.sort();
            println!(
                "{message}:\n{}",
                bulleted_list(&sorted.iter().map(|s| s.as_str()).collect::<Vec<_>>())
            );
        }

        if is_change_needed && !validate_options.allow_missing_change_files {
            eprintln!("ERROR: Change files are needed!");
            println!("{}", options.changehint);
            return Err(ValidationError {
                message: "Change files are needed".to_string(),
            }
            .into());
        }

        if options.disallow_deleted_change_files && are_change_files_deleted(options) {
            eprintln!("ERROR: Change files must not be deleted!");
            return Err(ValidationError {
                message: "Change files must not be deleted".to_string(),
            }
            .into());
        }

        changed_packages = Some(pkgs);
    }

    // Skip checkDependencies / bumpInMemory (not implemented)
    if validate_options.check_dependencies && !is_change_needed && !change_set.is_empty() {
        if options.verbose {
            println!(
                "(Skipping package dependency validation — not implemented in Rust port)"
            );
        }
    }

    println!();

    Ok(ValidationResult {
        is_change_needed,
        package_infos,
        package_groups,
        scoped_packages,
        change_set,
        changed_packages,
    })
}
